### PR TITLE
Fix edge cases in history (jax scalar) and add tests

### PR DESCRIPTION
**Pull request by PhilipVinc**
Previously History was not handling properly jax scalars.
Now it does, and a bunch of tests are added

### CODE REVIEW

File: netket/utils/history.py [hunk 1]
```diff
@@ -14,6 +14,29 @@
...
+
+    Checks whever it is a Python's number, or any numpy-API
+    scalar.
+    """
+    if isinstance(val, Number):
+        return True
+    elif np.isscalar(val):
+        return True
+    elif jnp.isscalar(val):
+        return True
```
Comment (femtobit) [inline netket/utils/history.py:32]: Hm, I was thinking that this could be simplified... but numpy / jax behavior is indeed different:
```python
In [11]: a
Out[11]: DeviceArray(1., dtype=float32)

In [12]: np.isscalar(a)
Out[12]: False

In [13]: jnp.isscalar(a)
Out[13]: True

In [14]: a = np.array(1.)

In [15]: np.isscalar(a)
Out[15]: False

In [16]: jnp.isscalar(a)
Out[16]: False
```

**Review by femtobit — APPROVED** [review]

File: netket/utils/history.py (hunk 1 above)
Comment (PhilipVinc) [reply netket/utils/history.py:32]: I think it's time to write a function using @dispatch to do this
